File: bindings/python/geometry.py
"""Converts between the client's pose types and WPILib's geometry types."""

from wpimath.geometry import Pose2d, Pose3d, Quaternion, Rotation2d, Rotation3d

from . import Pose2d as _Pose2d, Pose3d as _Pose3d


def convert(pose):
    """The WPILib pose for our fields, or our fields for a WPILib pose."""
    if isinstance(pose, _Pose3d):
        return Pose3d(
            pose.x,
            pose.y,
            pose.z,
            Rotation3d(Quaternion(pose.qw, pose.qx, pose.qy, pose.qz)),
        )
    if isinstance(pose, _Pose2d):
        return Pose2d(pose.x, pose.y, Rotation2d(pose.rotation))
    if isinstance(pose, Pose3d):
        rotation = pose.rotation().getQuaternion()
        return _Pose3d(
            pose.X(),
            pose.Y(),
            pose.Z(),
            rotation.W(),
            rotation.X(),
            rotation.Y(),
            rotation.Z(),
        )
    return _Pose2d(pose.X(), pose.Y(), pose.rotation().radians())

File: bindings/java/org/tarwyn/Geometry.java
package org.tarwyn;

import org.wpilib.math.geometry.Quaternion;
import org.wpilib.math.geometry.Rotation2d;
import org.wpilib.math.geometry.Rotation3d;

/** Converts between the client's pose types and WPILib's geometry types. */
public final class Geometry {
    private Geometry() {
    }

    /** The WPILib pose these fields describe. */
    public static org.wpilib.math.geometry.Pose2d convert(Pose2d pose) {
        return new org.wpilib.math.geometry.Pose2d(
            pose.x(), pose.y(), new Rotation2d(pose.rotation()));
    }

    /** The fields this client publishes for a WPILib pose. */
    public static Pose2d convert(org.wpilib.math.geometry.Pose2d pose) {
        return new Pose2d(pose.getX(), pose.getY(), pose.getRotation().getRadians());
    }

    /** The WPILib pose in space these fields describe. */
    public static org.wpilib.math.geometry.Pose3d convert(Pose3d pose) {
        return new org.wpilib.math.geometry.Pose3d(
            pose.x(), pose.y(), pose.z(),
            new Rotation3d(new Quaternion(pose.qw(), pose.qx(), pose.qy(), pose.qz())));
    }

    /** The fields this client publishes for a WPILib pose in space. */
    public static Pose3d convert(org.wpilib.math.geometry.Pose3d pose) {
        Quaternion rotation = pose.getRotation().getQuaternion();
        return new Pose3d(
            pose.getX(), pose.getY(), pose.getZ(),
            rotation.getW(), rotation.getX(), rotation.getY(), rotation.getZ());
    }
}
